contracts/staking/ServiceStakingBase.sol

```diff
@@ -562,26 +559,44 @@ abstract contract ServiceStakingBase is ERC721TokenReceiver, IErrorsRegistries {
             uint256[] memory eligibleServiceRewards, , ) = _calculateStakingRewards();
 
         // If there are eligible services, proceed with staking calculation and update rewards for the service Id
-        if (numServices > 0) {
+        for (uint256 i = 0; i < numServices; ++i) {
             // Get the service index in the eligible service set and calculate its latest reward
-            for (uint256 i = 0; i < eligibleServiceIds.length; ++i) {
-                if (eligibleServiceIds[i] == serviceId) {
-                    // If total allocated rewards are not enough, adjust the reward value
-                    if (totalRewards > lastAvailableRewards) {
-                        reward += (eligibleServiceRewards[i] * lastAvailableRewards) / totalRewards;
-                    } else {
-                        reward += eligibleServiceRewards[i];
-                    }
-                    break;
+            if (eligibleServiceIds[i] == serviceId) {
+                // If total allocated rewards are not enough, adjust the reward value
+                if (totalRewards > lastAvailableRewards) {
+                    reward += (eligibleServiceRewards[i] * lastAvailableRewards) / totalRewards;
+                } else {
+                    reward += eligibleServiceRewards[i];
                 }
+                break;
             }
         }
     }
 
+    /// @dev Gets staked service Ids.
+    /// @return serviceIds Staked service Ids.
+    function getServiceIds() public view returns (uint256[] memory serviceIds) {
+        // Get the number of service Ids
+        uint256 size = setServiceIds.length;
+        serviceIds = new uint256[](size);
+
+        // Record service Ids
+        for (uint256 i = 0; i < size; ++i) {
+            serviceIds[i] = setServiceIds[i];
+        }
+    }
+
     /// @dev Checks if the service is staked.
     /// @param serviceId.
-    /// @return True, if the service is staked.
-    function isServiceStaked(uint256 serviceId) external view returns (bool) {
-        return mapServiceInfo[serviceId].tsStart > 0;
+    /// @return isStaked True, if the service is staked.
+    function isServiceStaked(uint256 serviceId) external view returns (bool isStaked) {
+        isStaked = (mapServiceInfo[serviceId].tsStart > 0);
+    }
+
+    /// @dev Gets the next reward checkpoint timestamp.
+    /// @return tsNext Next reward checkpoint timestamp.
+    function getNextRewardCheckpointTimestamp() external view returns (uint256 tsNext) {
+        // Last checkpoint timestamp plus the liveness period
+        tsNext = tsCheckpoint + livenessPeriod;
```

Comment:
Getter of the next reward checkpoint timestamp - the agent can always compare with the `block.timestamp` and decide whether to call a `checkpoint()` or skip for now.